tenets.md

```diff
@@ -0,0 +1,431 @@
+that customers use to achieve their security requirements.
+If the ESDK is not correct,
+then it is impossible for customers
+to reason about their security requirements.
+
+Given a component that operates correctly,
+customers want the configuration that is correct for them.
+The above stages also apply to the use of the ESDK.
+
+Again furniture assembly is an example.
+By providing accurate pictures
+of each step and the finished product
+it is obvious after all the parts
+have been assembled that the table is complete.
+
+It is easier to verify
+that a system is correct
+when you're creating it,
+because systems tend toward complexity
+over time.
+As systems evolve, small, seemingly minor, changes
+can have unwanted emergent behavior.
+It is unlikely that your system
+will be _less_ complicated as time goes on.
+This is true both for the ESDK
+and for our customers using the ESDK.
+
+### Motivation
+
+The ESDK's priority is to
+create interfaces that are correct by construction.
+Our priority in identifying incorrectness MUST be
+
+1. Development
+1. Initialization
+1. Use
+
+Early detection of errors
+helps us deliver correct code to customers
+but it also helps customers build
+systems that have steady state stability.
+
+Our customers MUST be able to
+clearly express their requirements
+in the construction of their calls to our APIs.
+
+As a developer is integrating the ESDK into a project
+they MUST be able to [easily](#easy-to-use) transform
+a set of requirements into a implementation
+that is obviously correct.
+
+Similarly, reviewers or maintainers
+SHOULD be able to understand
+how the implementation satisfies the requirements.
+
+Being able to easily compose the ESDK parts
+to satisfy all of your requirements is not enough.
+We expect that the complete list of requirements
+to have conflicting security needs for different types of data.
+A single component that satisfies
+multiple conflicting requirements
+may be complete and expressive,
+but it is also confusing
+and is not [hard to misuse](#hard-to-misuse).
+
+Assembled components
+should do one thing well.
+Leaving the determination of
+what requirement is being satisfied
+to dynamic runtime behavior is suboptimal.
+
+## Clear Inputs
+
+### Summary
+
+Functions take arguments.
+What these arguments are
+and how they are intended to be used
+SHOULD be clear at the call site.
+This is a specific use of
+[correct by construction](#correct-by-construction)
+and [easy to use](#easy-to-use).
+
+Code is read much more often
+than it is written
+so we prioritize the expression of intent
+for those who come after us.
+By structuring clarity at the call site
+we maintain the readability of our codebase.
+By structuring our API
+to encourage this same kind of clarity
+we help customers express their intent
+when using our tools.
+
+Examples of features used to express clear inputs
+at the call site are
+named parameters, enums, and options mappings.
```

Comment:
Another example is precondition declarations, which detail expected properties of the inputs.